src/message.rs

```diff
@@ -0,0 +1,211 @@
+use std::borrow::Cow;
+use std::collections::BTreeMap;
+use std::convert::TryFrom;
+use std::fmt;
+use std::fmt::Write;
+use std::option::Option;
+
+use super::error::ParseError;
+
+use crate::escaped::{escape_char, unescape_char};
+
+#[derive(Debug, PartialEq, Default)]
+pub struct Message<'a> {
+    pub tags: BTreeMap<&'a str, Cow<'a, str>>,
+    pub prefix: Option<&'a str>,
+    pub command: &'a str,
+    pub params: Vec<&'a str>,
+}
+
+fn parse_tags<'a>(input: &'a str) -> Result<BTreeMap<&'a str, Cow<'a, str>>, ParseError> {
+    let mut tags = BTreeMap::new();
+
+    for tag_data in input.split(';') {
+        let (tag_name, raw_tag_value) = if let Some(loc) = tag_data.find('=') {
+            (&tag_data[..loc], tag_data.get(loc + 1..).unwrap_or(""))
+        } else {
+            // If there's no equals sign, we need to default to the empty
+            // string/
+            (tag_data, "")
+        };
+
+        // If the value doesn't contain any escaped characters, we can return
+        // the string as-is.
+        if !raw_tag_value.contains('\\') {
+            tags.insert(tag_name, Cow::Borrowed(raw_tag_value));
+            continue;
+        }
+
+        let mut tag_value = String::new();
+        let mut tag_value_chars = raw_tag_value.chars();
+        while let Some(c) = tag_value_chars.next() {
+            if c == '\\' {
+                if let Some(escaped_char) = tag_value_chars.next() {
+                    tag_value.push(unescape_char(escaped_char));
+                }
+            } else {
+                tag_value.push(c);
+            }
+        }
+
+        tags.insert(tag_name, Cow::Owned(tag_value));
+    }
+
+    Ok(tags)
+}
+
+impl<'a> TryFrom<&'a str> for Message<'a> {
+    type Error = ParseError;
+
+    fn try_from(input: &'a str) -> Result<Self, Self::Error> {
+        // We want a mutable input so we can jump through it as we parse the
+        // message.
+        let mut input = input;
+
+        // Possibly chop off the ending \r\n where either of those characters is
+        // optional.
+        if input.ends_with('\n') {
+            input = &input[..input.len() - 1];
+        }
+        if input.ends_with('\r') {
+            input = &input[..input.len() - 1];
+        }
```

Comment:
This is slightly different behavior. Note that you can also do input.trim_end_matches("\r\n").

I saw this but didn't do it because it potentially trims multiple \n's and \r's from the end of the line.